dangerzone/gui/updater.py

```diff
@@ -0,0 +1,225 @@
+        # TODO: Render HTML links properly
+        # FIXME: Handle the case where a user clicks on "X", instead of explicity making
+        # a choice. We should probably ask them again on the next run.
+        check = Alert(
+            self.dangerzone,
+            message=MSG_CONFIRM_UPDATE_CHECKS,
+            ok_text="Yes",
+            cancel_text="No",
+        ).launch()
+        return bool(check)
+
+    def should_check_for_updates(self) -> bool:
+        """Determine if we can check for updates based on settings and user prefs."""
+        log.debug("Checking platform type")
+        # TODO: Disable updates for Homebrew installations.
+        # TODO: Add a flag that will force-check for updates, so that we can test it on
+        # Linux. Once we have this flag, uncomment the following lines.
+        # if platform.system() == "Linux" and os.environ():
+        #     log.debug("Running on Linux, disabling updates")
+        #     self.check = False
+        #     return False
+
+        log.debug("Checking if first run of Dangerzone")
+        if self.first_run:
+            log.debug("Dangerzone is running for the first time, updates are stalled")
+            self.first_run = False
+            return False
+
+        log.debug("Checking if user has already expressed their preference")
+        if self.check is None:
+            log.debug("User has not been asked yet for update checks")
+            self.check = self.prompt_for_checks()
+            # TODO: Should we always check for updates in the third run, or can we check
+            # them immediately, if the user wants to enable them?
+            return False
+        elif not self.check:
+            log.debug("User has expressed that they don't want to check for updates")
+            return False
+
+        return True
+
+    def can_update(self, cur_version: str, latest_version: str) -> bool:
+        if cur_version == latest_version:
+            return False
+        elif cur_version > latest_version:
+            # FIXME: This is a sanity check, but we should improve its wording.
+            raise Exception("Received version is older than the latest version")
+        else:
+            return True
+
+    def create_update_report(self, version: str, changelog: str) -> dict:
+        return {
+            "version": version,
+            "changelog": changelog,
+            "error": None,
+        }
+
+    def get_latest_info(self) -> dict:
+        """Get the latest release info from GitHub.
+
+        Also, render the changelog from Markdown format to HTML, so that we can show it
+        to the users.
+        """
+        try:
+            # TODO: Set timeout.
+            info = requests.get(self.GH_RELEASE_URL).json()
+        except Exception as e:
+            # TODO:: Handle errors.
+            raise
+
+        version = info["tag_name"].lstrip("v")
+        changelog = markdown.markdown(info["body"])
```

Comment:
We actually have an outer `try...catch` block, which would catch this error and report the traceback. The inner `try...catch` block is a placeholder in case we want to explicitly translate a weird `requests` / `urllib` error.

Reply:
You are right. I wish python made it easier to see where errors are caught.